higher-order-functions/exercise.js

```diff
@@ -27,31 +44,92 @@ function forEach(array, callback) {}
 //--------------------------------------------------
 
 //Extension 1
-function mapWith(array, callback) {}
+function multiplyByTwo(n) {
+    return n*2;
+}
+function forEach(array, callback) {
+    for (let i = 0; i < array.length; i++) {
+       callback(array[i])
+    }
+}
+function mapWith(array,callback) {
+    var newArray = [];	
+    forEach(array, (num)=> {
+        newArray.push(callback(num))
+    })
+    return newArray;
+ };
 
 //Extension 2
-function reduce(array, callback, initialValue) {}
+function reduce(array, callback, initialValue) {
+    var accumelator = initialValue;
+    for(let i = 0; i < array.length; i++) {
+      accumelator += callback(accumelator,array[i]);
+    } return accumelator
+}
 
 //Extension 3
-function intersection(arrays) {}
+function intersection(...arrays) {
+	let result = [];
+	arrays.reduce((acc, cv) => {
+		acc.forEach(element => {
+			if (cv.includes(element) && !result.includes(element)) {
+				result.push(element);}
+		});
+		return result;
+	});
+	return result;
+}
 
 // console.log(intersection([5, 10, 15, 20], [15, 88, 1, 5, 7], [1, 10, 15, 5, 20]));
 // should log: [5, 15]
-
+ 
 //Extension 4
-function union(arrays) {}
+function union(...arrays) {
+	let result = [];
+	arrays.reduce((acc, cv) => {
+		cv.forEach(item => {
+			if (!acc.includes(item)) {
+				result.push(item);
+			}
+		});
+		return result;
+	}, []);
+	return result;
+}
 
 // console.log(union([5, 10, 15], [15, 88, 1, 5, 7], [100, 15, 10, 1, 5]));
 // should log: [5, 10, 15, 88, 1, 7, 100]
 
 //Extension 5
-function objOfMatches(array1, array2, callback) {}
+function objOfMatches(array1, array2, callback) {
+	let obj = {};
+	array1.reduce((accumulator, currentValue) => {
+		let output = callback(currentValue);
+		let index = array1.indexOf(currentValue);
+		if (output == array2[index]) {
+			accumulator[currentValue] = array2[index];
+		}
+		obj = accumulator;
+		return accumulator;
+	}, {});
+	return obj;
+}
 
 // console.log(objOfMatches(['hi', 'howdy', 'bye', 'later', 'hello'], ['HI', 'Howdy', 'BYE', 'LATER', 'hello'], function(str) { return str.toUpperCase(); }));
 // should log: { hi: 'HI', bye: 'BYE', later: 'LATER' }
 
 //Extension 6
-function multiMap(arrVals, arrCallbacks) {}
+function multiMap(arrVals, arrCallbacks) {
+	let result = {};
+	arrVals.reduce((accumulator, currentValue) => {
+		accumulator[currentValue] = arrCallbacks.map(item => item(currentValue));
+		result = accumulator;
+		return accumulator;
+	}, {});
+	return result;
+}
+
```

Comment:
You don't need to save the result in `result` variable. You can directly return like below
```js
function multiMap(arrVals, arrCallbacks) {
	return arrVals.reduce((accumulator, currentValue) => {
		accumulator[currentValue] = arrCallbacks.map(item => item(currentValue));
		return accumulator;
	}, {});
}
```